cfg/builder/builder_entry.cc

```diff
@@ -36,13 +36,82 @@ unique_ptr<CFG> CFGBuilder::buildFor(core::Context ctx, ast::MethodDef &md) {
                                          selfClaz.data(ctx)->enclosingClass(ctx).data(ctx)->selfType(ctx),
                                          core::Names::cast()));
 
+        BasicBlock *presentCont = entry;
+        BasicBlock *defaultCont = nullptr;
+
+        auto &argInfos = md.symbol.data(ctx)->arguments();
+        bool isAbstract = md.symbol.data(ctx)->isAbstract();
+        bool seenKeyword = false;
         int i = -1;
         for (auto &argExpr : md.args) {
             i++;
             auto *a = ast::MK::arg2Local(argExpr);
-            synthesizeExpr(entry, res->enterLocal(a->localVariable), a->loc, make_unique<LoadArg>(md.symbol, i));
+            auto local = res->enterLocal(a->localVariable);
+            auto &info = argInfos[i];
+
+            seenKeyword = seenKeyword || info.flags.isKeyword;
+
+            // If defaultCont is non-null, that means that the previous argument had a default. If the current argument
+            // also has a default, and is not a keyword, block or repeated arg, then we can continue by extending that
+            // fall-through case. However if any of those conditions fail, we must merge the two paths back together,
+            // and break out of the fast-path for defaulting.
+            if (defaultCont && (seenKeyword || info.flags.isBlock || info.flags.isRepeated || !info.flags.isDefault)) {
+                auto *join = res->freshBlock(cctx.loops, presentCont->rubyBlockId);
+                unconditionalJump(presentCont, join, *res, core::LocOffsets::none());
+                unconditionalJump(defaultCont, join, *res, core::LocOffsets::none());
+                presentCont = join;
+                defaultCont = nullptr;
+            }
+
+            // Ignore defaults for abstract methods
+            if (!isAbstract) {
+                // Only emit conditional arg loading if the arg has a default
+                if (auto *opt = ast::cast_tree<ast::OptionalArg>(argExpr)) {
+                    auto defLoc = opt->default_->loc;
+
+                    auto *presentNext = res->freshBlock(cctx.loops, presentCont->rubyBlockId);
+                    auto *defaultNext = res->freshBlock(cctx.loops, presentCont->rubyBlockId);
+
+                    auto present = cctx.newTemporary(core::Names::argPresent());
+                    synthesizeExpr(presentCont, present, core::LocOffsets::none(),
```

Comment:
Is this `LocOffsets::none()` load bearing?

Reply:
No, it could be `defLoc`. Given that the `ArgPresent` result doesn't correspond to anything in the source program, I though it was reasonable to leave it with no location information. I'd be happy to set it to the location of the whole arg, though.

Reply:
👍 i think it's a safer default to leave it set to something (you never know when ??? is going to show up in an error message after sorbet runs for a long time)